File: cppFolder/coinBoard.cpp
#include <iostream>
#include <queue>
#include <map>
#include <vector>
#include <algorithm>
#include <utility>
#include <clocale>

void printMask(int mask, int N, int M) {
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < M; ++j) {
            int bit_pos = i * M + j;
            int bit_value = (mask >> bit_pos) & 1;
            std::cout << bit_value << " ";
        }
        std::cout << "\n";
    }
    std::cout << "\n";
}

std::pair<int, int> getTargetMasks(int N, int M) {
    int target1 = 0;
    int target2 = 0;

    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < M; ++j) {
            int bit_pos = i * M + j;
            if ((i + j) % 2 == 0) {
                target1 |= (1 << bit_pos);
            } else {
                target2 |= (1 << bit_pos);
            }
        }
    }
    return {target1, target2};
}

std::vector<int> solveChessboard(int N, int M, int start_mask) {
    auto [target1, target2] = getTargetMasks(N, M);

    if (start_mask == target1 || start_mask == target2) {
        return {start_mask};
    }

    std::queue<int> q;
    std::map<int, int> dist;   
    std::map<int, int> parent; 

    q.push(start_mask);
    dist[start_mask] = 0;
    parent[start_mask] = -1;

    int end_mask = -1;

    while (!q.empty()) {
        int curr = q.front();
        q.pop();
        int d = dist[curr];

        if (curr == target1 || curr == target2) {
            end_mask = curr;
            break;
        }

        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < M; ++j) {
                int pos1 = i * M + j;

                if (j + 1 < M) {
                    int pos2 = i * M + (j + 1);
                    int next_mask = curr ^ (1 << pos1) ^ (1 << pos2);
                    if (dist.find(next_mask) == dist.end()) {
                        dist[next_mask] = d + 1;
                        parent[next_mask] = curr;
                        q.push(next_mask);
                    }
                }

                if (i + 1 < N) {
                    int pos2 = (i + 1) * M + j;
                    int next_mask = curr ^ (1 << pos1) ^ (1 << pos2);
                    if (dist.find(next_mask) == dist.end()) {
                        dist[next_mask] = d + 1;
                        parent[next_mask] = curr;
                        q.push(next_mask);
                    }
                }
            }
        }
    }

    if (end_mask == -1) {
        return {};
    }

    std::vector<int> path;
    int curr_node = end_mask;
    while (curr_node != -1) {
        path.push_back(curr_node);
        curr_node = parent[curr_node];
    }
    std::reverse(path.begin(), path.end());

    return path;
}

void runChessboardSolver() {
    int N, M;
    std::cout << "Введите размеры доски N и M: ";
    if (!(std::cin >> N >> M)) return;

    std::cout << "Введите матрицу состояний (0 - орел, 1 - решка):\n";
    int start_mask = 0;
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < M; ++j) {
            int val;
            std::cin >> val;
            if (val == 1) {
                start_mask |= (1 << (i * M + j));
            }
        }
    }

    std::vector<int> path = solveChessboard(N, M, start_mask);

    if (path.empty()) {
        std::cout << "\nРешения не существует\n";
        return;
    }

    if (path.size() == 1) {
        std::cout << "\nНачальное состояние уже является целевым!\n";
        printMask(path[0], N, M);
        std::cout << "Минимальное число ходов: 0\n";
        return;
    }

    std::cout << "\nПромежуточные состояния доски\n\n";
    for (size_t i = 0; i < path.size(); ++i) {
        if (i == 0) {
            std::cout << "Начальное состояние:\n";
        } else {
            std::cout << "Ход №" << i << ":\n";
        }
        printMask(path[i], N, M);
    }

    std::cout << "Минимальное число ходов: " << path.size() - 1 << "\n";
}

File: goFolder/coinBoard/coinBoard.go
package coinBoard

import "fmt"

func printMask(mask, N, M int) {
	for i := 0; i < N; i++ {
		for j := 0; j < M; j++ {
			bitPos := i*M + j
			bitValue := (mask >> bitPos) & 1
			fmt.Printf("%d ", bitValue)
		}
		fmt.Println()
	}
	fmt.Println()
}

func getTargetMasks(N, M int) (int, int) {
	target1 := 0
	target2 := 0
	for i := 0; i < N; i++ {
		for j := 0; j < M; j++ {
			bitPos := i*M + j
			if (i+j)%2 == 0 {
				target1 |= (1 << bitPos)
			} else {
				target2 |= (1 << bitPos)
			}
		}
	}
	return target1, target2
}

func solveChessboard(N, M, startMask int) []int {
	target1, target2 := getTargetMasks(N, M)
	if startMask == target1 || startMask == target2 {
		return []int{startMask}
	}
	var q []int
	dist := make(map[int]int)
	parent := make(map[int]int)
	q = append(q, startMask)
	dist[startMask] = 0
	parent[startMask] = -1
	endMask := -1
	for len(q) > 0 {
		curr := q[0]
		q = q[1:]
		d := dist[curr]
		if curr == target1 || curr == target2 {
			endMask = curr
			break
		}
		for i := 0; i < N; i++ {
			for j := 0; j < M; j++ {
				pos1 := i*M + j
				if j+1 < M {
					pos2 := i*M + (j + 1)
					nextMask := curr ^ (1 << pos1) ^ (1 << pos2)
					if _, exists := dist[nextMask]; !exists {
						dist[nextMask] = d + 1
						parent[nextMask] = curr
						q = append(q, nextMask)
					}
				}
				if i+1 < N {
					pos2 := (i+1)*M + j
					nextMask := curr ^ (1 << pos1) ^ (1 << pos2)
					if _, exists := dist[nextMask]; !exists {
						dist[nextMask] = d + 1
						parent[nextMask] = curr
						q = append(q, nextMask)
					}
				}
			}
		}
	}
	if endMask == -1 {
		return nil
	}
	var path []int
	currNode := endMask
	for currNode != -1 {
		path = append(path, currNode)
		currNode = parent[currNode]
	}
	for i, j := 0, len(path)-1; i < j; i, j = i+1, j-1 {
		path[i], path[j] = path[j], path[i]
	}
	return path
}

func RunChessboardSolver() {
	var N, M int
	fmt.Print("Введите размеры доски N и M: ")
	if _, err := fmt.Scan(&N, &M); err != nil {
		return
	}
	fmt.Println("Введите матрицу состояний (0 - орел, 1 - решка):")
	startMask := 0
	for i := 0; i < N; i++ {
		for j := 0; j < M; j++ {
			var val int
			fmt.Scan(&val)
			if val == 1 {
				startMask |= (1 << (i*M + j))
			}
		}
	}
	path := solveChessboard(N, M, startMask)
	if path == nil {
		fmt.Println("\nРешения не существует")
		return
	}
	if len(path) == 1 {
		fmt.Println("\nНачальное состояние уже является целевым!")
		printMask(path[0], N, M)
		fmt.Println("Минимальное число ходов: 0")
		return
	}
	fmt.Println("\nПромежуточные состояния доски")
	for i := 0; i < len(path); i++ {
		if i == 0 {
			fmt.Println("Начальное состояние:")
		} else {
			fmt.Printf("Ход №%d:\n", i)
		}
		printMask(path[i], N, M)
	}
	fmt.Printf("Минимальное число ходов: %d\n", len(path)-1)
}
